Editorial: Convert remaining ZonedDateTime AOs to structured headers

Lubrsi opened the pull request.

ptomato commented on spec/zoneddatetime.html:
```diff
@@ -1115,11 +1137,19 @@ <h1>InterpretISODateTimeOffset ( _year_, _month_, _day_, _hour_, _minute_, _seco
+      <dl class="header">
+        <dt>description</dt>
+        <dd>
+          It returns its argument _item_ if it is already a Temporal.ZonedDateTime instance, converts _item_ to a new Temporal.PlainDate instance if possible, and throws otherwise.
```
```suggestion
          It returns its argument _item_ if it is already a Temporal.ZonedDateTime instance, converts _item_ to a new Temporal.ZonedDateTime instance if possible, and throws otherwise.
```
Optional, but we may as well fix this while you're in here anyway :smile:

Lubrsi replied:
Done :+1:

ljharb commented on spec/zoneddatetime.html:
```diff
@@ -1078,18 +1078,40 @@ <h1>Properties of Temporal.ZonedDateTime Instances</h1>
+          _millisecond_: an integer,
+          _microsecond_: an integer,
+          _nanosecond_: an integer,
+          _offsetBehaviour_: one of ~option~, ~exact~ or ~wall~,
```
```suggestion
          _offsetBehaviour_: one of ~option~, ~exact~, or ~wall~,
```

Lubrsi replied:
Done :+1:

ljharb commented on spec/zoneddatetime.html:
```diff
@@ -1078,18 +1078,40 @@ <h1>Properties of Temporal.ZonedDateTime Instances</h1>
+          _offsetBehaviour_: one of ~option~, ~exact~ or ~wall~,
+          _offsetNanoseconds_: an integer,
+          _timeZone_: an Object,
+          _disambiguation_: one of *"earlier"*, *"later"*, *"compatible"*, or *"reject"*,
+          _offsetOption_: one of *"ignore"*, *"use"*, *"prefer"*, or *"reject"*,
```
should these be converted from strings to spec enums earlier in the process? (unrelated to this PR)

ljharb commented on spec/zoneddatetime.html:
```diff
@@ -1262,14 +1309,24 @@ <h1>
+    <emu-clause id="sec-temporal-differencezoneddatetime" type="abstract operation">
+      <h1>
+        DifferenceZonedDateTime (
+          _ns1_: a BigInt value,
+          _ns2_: a BigInt value,
```
do we say `a BigInt value` here, as opposed to just `a BigInt`?

Lubrsi replied:
There's a couple already in use, but I don't mind changing this:
https://github.com/tc39/proposal-temporal/blob/f47ebd8475fa016d60e6e8f44a03b5ea4f14de5f/spec/zoneddatetime.html#L1225
https://github.com/tc39/proposal-temporal/blob/f47ebd8475fa016d60e6e8f44a03b5ea4f14de5f/spec/instant.html#L570

ljharb replied:
as long as ecmarkup supports it, the more concise form seems better :-)

Lubrsi replied:
Done :+1:

ljharb commented on spec/zoneddatetime.html:
```diff
@@ -1262,14 +1309,24 @@ <h1>
+          _calendar_: an Object,
+          _largestUnit_: a String,
+          optional _options_: an Object,
+        )
```
what type does it return?

Lubrsi replied:
It returns a `Duration Record`. Would it be fine to just do `: Duration Record` or is there another way of doing this?

ljharb replied:
`: a Duration Record` seems great!

Lubrsi replied:
Done :+1:

ljharb commented on spec/zoneddatetime.html:
```diff
@@ -1187,13 +1222,26 @@ <h1>CreateTemporalZonedDateTime ( _epochNanoseconds_, _timeZone_, _calendar_ [ ,
+          optional _increment_: an integer,
+          optional _unit_: a String,
+          optional _roundingMode_: a String,
+        )
```
```suggestion
        ): a String
```

Lubrsi replied:
When I do this, I get a bunch of errors:
```
Error: spec/zoneddatetime.html:1252:38: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1253:36: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1257:32: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1262:34: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1264:32: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:839:1: TemporalZonedDateTimeToString does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:855:1: TemporalZonedDateTimeToString does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:867:1: TemporalZonedDateTimeToString does not return a Completion Record, but is consumed as if it does
```

ljharb replied:
ah, in that case:
```suggestion
        ): a normal completion containing a String
```

Lubrsi replied:
Done :+1:

ljharb commented on spec/zoneddatetime.html:
```diff
@@ -1168,16 +1198,21 @@ <h1>ToTemporalZonedDateTime ( _item_ [ , _options_ ] )</h1>
+          _timeZone_: an Object,
+          _calendar_: an Object,
+          optional _newTarget_: a function object,
+        )
```
```suggestion
        ): an Object
```

Lubrsi replied:
Ditto:
```
Error: spec/zoneddatetime.html:1217:28: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/temporal.html:260:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/plaindate.html:526:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/plaintime.html:419:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/plaindatetime.html:648:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:40:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:68:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:586:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:609:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:629:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:643:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:657:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:675:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:693:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:804:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:895:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:1197:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:1339:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/duration.html:1558:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/instant.html:432:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/instant.html:451:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/abstractops.html:517:1: CreateTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
```

ljharb replied:
```suggestion
        ): a normal completion containing an Object
```

Lubrsi replied:
Done :+1:

ljharb commented on spec/zoneddatetime.html:
```diff
@@ -1115,11 +1137,19 @@ <h1>InterpretISODateTimeOffset ( _year_, _month_, _day_, _hour_, _minute_, _seco
+        ToTemporalZonedDateTime (
+          _item_: an ECMAScript language value,
+          optional _options_: an Object,
+        )
```
```suggestion
        ): an ECMAScript language value
```

Lubrsi replied:
Ditto:
```
Error: spec/zoneddatetime.html:1161:32: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1162:34: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1165:30: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1166:32: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1167:32: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1168:36: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1172:38: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1173:30: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1175:22: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1176:30: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1177:30: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1181:69: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1189:32: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1193:41: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1194:36: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1195:34: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1196:38: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:69:1: ToTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:80:1: ToTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:81:1: ToTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:706:1: ToTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:741:1: ToTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:817:1: ToTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
```

ljharb replied:
```suggestion
        ): a normal completion containing an ECMAScript language value
```

Lubrsi replied:
Done :+1:

ljharb commented on spec/zoneddatetime.html:
```diff
@@ -1078,18 +1078,40 @@ <h1>Properties of Temporal.ZonedDateTime Instances</h1>
+          _disambiguation_: one of *"earlier"*, *"later"*, *"compatible"*, or *"reject"*,
+          _offsetOption_: one of *"ignore"*, *"use"*, *"prefer"*, or *"reject"*,
+          _matchBehaviour_: one of ~match exactly~ or ~match minutes~,
+        )
```
```suggestion
        ): an integer
```

Lubrsi replied:
Ditto:
```
Error: spec/zoneddatetime.html:1116:30: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1118:31: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1125:38: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1127:44: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1134:45: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1135:29: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1161:32: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1162:34: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1165:30: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1166:32: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1167:32: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1168:36: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1172:38: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1173:30: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1175:22: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1176:30: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1177:30: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1181:69: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1189:32: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1193:41: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1194:36: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1195:34: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:1196:38: this would return a Completion Record, but the containing AO is declared not to return a Completion Record
Error: spec/zoneddatetime.html:69:1: ToTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:80:1: ToTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:81:1: ToTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:585:1: InterpretISODateTimeOffset does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:706:1: ToTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:741:1: ToTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:803:1: InterpretISODateTimeOffset does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:817:1: ToTemporalZonedDateTime does not return a Completion Record, but is consumed as if it does
Error: spec/zoneddatetime.html:1196:1: InterpretISODateTimeOffset does not return a Completion Record, but is consumed as if it does
Error: spec/abstractops.html:516:1: InterpretISODateTimeOffset does not return a Completion Record, but is consumed as if it does
```

ljharb replied:
```suggestion
        ): a normal completion containing an integer
```

Lubrsi replied:
Done :+1:

ptomato commented on spec/zoneddatetime.html:
```diff
@@ -1187,13 +1222,26 @@ <h1>CreateTemporalZonedDateTime ( _epochNanoseconds_, _timeZone_, _calendar_ [ ,
+      <h1>
+        TemporalZonedDateTimeToString (
+          _zonedDateTime_: a Temporal.ZonedDateTime,
+          _precision_: a String or integer,
```
This is probably OK, but we could also make it even more specific, e.g. "one of **"auto"**, **"minute"**, or an integer between 0 and 9 inclusive".
Ditto for the other options below, e.g. `showCalendar` could be "one of **"auto"**, **"always"**, or **"never"**".

Lubrsi replied:
Done :+1:

ptomato reviewed (approved):
Thanks!! Really helpful.

By the way, I think it's OK to omit the return types for now, as you have done here; that's what I've been doing as well. They will require  a bunch of refactoring of the spec in order to correctly handle completion records. I will plan to add return types to the structured headers later.

Lubrsi commented:
I've gone ahead and added the return values for all the AOs :+1: 
I've assumed that any fallible operation inside an AO makes it also possible to return an abrupt completion, so I've added that in too.

Lubrsi commented on spec/zoneddatetime.html:
```diff
@@ -1115,11 +1137,19 @@ <h1>InterpretISODateTimeOffset ( _year_, _month_, _day_, _hour_, _minute_, _seco
+        ToTemporalZonedDateTime (
+          _item_: an ECMAScript language value,
+          optional _options_: an Object,
+        ): either a normal completion containing an Object or an abrupt completion
```
I originally thought about doing this but just went with `Object` to be safe.
Done :+1:

Lubrsi commented on spec/zoneddatetime.html:
```diff
@@ -1168,16 +1198,21 @@ <h1>ToTemporalZonedDateTime ( _item_ [ , _options_ ] )</h1>
+          _timeZone_: an Object,
+          _calendar_: an Object,
+          optional _newTarget_: a function object,
+        ): either a normal completion containing an Object or an abrupt completion
```
Applied the suggestion :+1:

ptomato commented:
Thanks!
